### PR TITLE
fix(history): Prevent invalid history access

### PR DESCRIPTION
**Pull request by sphaerophoria**
* When the DB schema was too new we were accessing history anyways. This
has potential to just completely corrupt the DB
* When history was disabled there was a chance we would attempt to write
to history anyways. Added more checks in this area

Tested by bumping my DB version up to two, committing that to history, then dropping it back down to 1

Stop gap solution for #5683. I'd like to leave that open until we implement a gui popup but getting this in early prevents a lot of invalid history accesses that could cause corruption

---
This change is [Reviewable](https://reviewable.io/reviews/qtox/qtox/5828)

### CODE REVIEW

**Review by sudden6 — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5828)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/history.cpp, line 681 at r1](https://reviewable.io/reviews/qtox/qtox/5828#-Loj4X2e0pZJcYqPOCY8:-Loj4X2f8Hq8VBgUJ02g:b-vnvj3e) ([raw file](https://github.com/qtox/qtox/blob/bc609554a8c13c7a63d683dccb3663298dde309c/src/persistence/history.cpp#L681)):*
> ```cpp
> {
>     if (historyAccessBlocked()) {
>         return QDateTime::currentDateTime();
> ```

why is the current date time returned here and not the default constructed one?
___
*[src/persistence/history.cpp, line 846 at r1](https://reviewable.io/reviews/qtox/qtox/5828#-Loj4eiGB12j1CeWfEv_:-Loj4eiHD7SfgmwEpEv1:by6r3lf) ([raw file](https://github.com/qtox/qtox/blob/bc609554a8c13c7a63d683dccb3663298dde309c/src/persistence/history.cpp#L846)):*
> ```cpp
> bool History::historyAccessBlocked()
> {
>     if (!Settings::getInstance().getEnableLogging()) {
> ```

Do not use the settings singleton, we want to get rid of it in the long term. Is this check even needed or is it just a to be extra sure thing?

**Review by sphaerophoria — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5828)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/history.cpp, line 681 at r1](https://reviewable.io/reviews/qtox/qtox/5828#-Loj4X2e0pZJcYqPOCY8:-LokWWgF9I1losytaqP8:b-alzru1) ([raw file](https://github.com/qtox/qtox/blob/bc609554a8c13c7a63d683dccb3663298dde309c/src/persistence/history.cpp#L681)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

why is the current date time returned here and not the default constructed one?
</blockquote></details>

I wasn't sure if callers handled a null QDateTime correctly. This call is used in order to see where in history to load from when doing a search so I figured currentDateTime made most sense and had the least chance of going wrong.

You're right though it does make more sense from an API perspective to make this a null one. I'll double check if callers can handle this correctly
___
*[src/persistence/history.cpp, line 846 at r1](https://reviewable.io/reviews/qtox/qtox/5828#-Loj4eiGB12j1CeWfEv_:-LokW-9y2i_RhaUd-UAb:b-2qet6t) ([raw file](https://github.com/qtox/qtox/blob/bc609554a8c13c7a63d683dccb3663298dde309c/src/persistence/history.cpp#L846)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

Do not use the settings singleton, we want to get rid of it in the long term. Is this check even needed or is it just a to be extra sure thing?
</blockquote></details>

I understand that we don't want to use it, but I just moved an existing call. On one out of many public apis we were checking if we should allow people to write to the DB. It seems like if it's valid on one public API it's valid on them all.

**Review by anthonybilinski — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5828)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/history.cpp, line 846 at r1](https://reviewable.io/reviews/qtox/qtox/5828#-Loj4eiGB12j1CeWfEv_:-LorOsa_0vhdkwIIsc8W:bszgxph) ([raw file](https://github.com/qtox/qtox/blob/bc609554a8c13c7a63d683dccb3663298dde309c/src/persistence/history.cpp#L846)):*
<details><summary><i>Previously, sphaerophoria wrote…</i></summary><blockquote>

I understand that we don't want to use it, but I just moved an existing call. On one out of many public apis we were checking if we should allow people to write to the DB. It seems like if it's valid on one public API it's valid on them all. 
</blockquote></details>

re: Is this check even needed or is it just a to be extra sure thing?
I added it in https://github.com/qTox/qTox/issues/5036 as a double check.  I agree that removing an existing singleton is out of scope for a bug fix on a release branch - even though it's a good idea long term.
___
*[src/persistence/history.cpp, line 847 at r1](https://reviewable.io/reviews/qtox/qtox/5828#-LorQ6GR4FManMO3ZI03:-LorQ6GR4FManMO3ZI04:b4zrftx) ([raw file](https://github.com/qtox/qtox/blob/bc609554a8c13c7a63d683dccb3663298dde309c/src/persistence/history.cpp#L847)):*
> ```cpp
> {
>     if (!Settings::getInstance().getEnableLogging()) {
>         qWarning() << "Blocked history access while history is disabled";
> ```

I know that this is an existing Warning log, but could you bump it to Critical and add an assert? It's effectively a security issue that is narrowly being avoided if it gets to this point.

**Review by sudden6 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5828)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/history.cpp, line 846 at r1](https://reviewable.io/reviews/qtox/qtox/5828#-Loj4eiGB12j1CeWfEv_:-LpDv-8K9hYmcnJVv2xn:bxr6e8u) ([raw file](https://github.com/qtox/qtox/blob/bc609554a8c13c7a63d683dccb3663298dde309c/src/persistence/history.cpp#L846)):*
<details><summary><i>Previously, anthonybilinski (Anthony Bilinski) wrote…</i></summary><blockquote>

re: Is this check even needed or is it just a to be extra sure thing?
I added it in https://github.com/qTox/qTox/issues/5036 as a double check.  I agree that removing an existing singleton is out of scope for a bug fix on a release branch - even though it's a good idea long term.
</blockquote></details>

To clarify, I don't want you to remove that singleton. AFAICT this class does not use the Settings singleton ATM, so I just don't want you to introduce a dependency on it, because it will make removing the singleton a lot harder in the long run if we keep introducing new uses to it.

**Review by sphaerophoria — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5828)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/history.cpp, line 496 at r1](https://reviewable.io/reviews/qtox/qtox/5828#-LpK4_z24UT7M3gswfYq:-LpK4_z24UT7M3gswfYr:b-80h7ta) ([raw file](https://github.com/qtox/qtox/blob/bd034e130d1b0630f272d9b36a33b0e16bc44d9e/src/persistence/history.cpp#L496)):*
> ```cpp
>                             const std::function<void(RowId)>& insertIdCallback)
> {
>     if (!Settings::getInstance().getEnableLogging()) {
> ```

@sudden6 all I did was move this line into a new function.

**Review by sudden6 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5828)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/history.cpp, line 496 at r1](https://reviewable.io/reviews/qtox/qtox/5828#-LpK4_z24UT7M3gswfYq:-LpO4C1iAd7TW7NNuNTz:b-aqesrp) ([raw file](https://github.com/qtox/qtox/blob/bd034e130d1b0630f272d9b36a33b0e16bc44d9e/src/persistence/history.cpp#L496)):*
<details><summary><i>Previously, sphaerophoria wrote…</i></summary><blockquote>

@sudden6 all I did was move this line into a new function.
</blockquote></details>

damn missed that, then I retract my comment adding the settings singleton here.
___
*[src/persistence/history.cpp, line 681 at r1](https://reviewable.io/reviews/qtox/qtox/5828#-Loj4X2e0pZJcYqPOCY8:-LpO4Ro4BPi6pV7OHPKo:bxz2rzz) ([raw file](https://github.com/qtox/qtox/blob/bc609554a8c13c7a63d683dccb3663298dde309c/src/persistence/history.cpp#L681)):*
<details><summary><i>Previously, sphaerophoria wrote…</i></summary><blockquote>

I wasn't sure if callers handled a null QDateTime correctly. This call is used in order to see where in history to load from when doing a search so I figured currentDateTime made most sense and had the least chance of going wrong.

You're right though it does make more sense from an API perspective to make this a null one. I'll double check if callers can handle this correctly
</blockquote></details>

If you decide to not use the default constructed value, please add a comment describing your reasoning.

**Review by sphaerophoria — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5828)* status: 0 of 1 LGTMs obtained

___
*[src/persistence/history.cpp, line 681 at r1](https://reviewable.io/reviews/qtox/qtox/5828#-Loj4X2e0pZJcYqPOCY8:-LtpsWaZ648BxnAck1_t:b-oleq0t) ([raw file](https://github.com/qtox/qtox/blob/bc609554a8c13c7a63d683dccb3663298dde309c/src/persistence/history.cpp#L681)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

If you decide to not use the default constructed value, please add a comment describing your reasoning.
</blockquote></details>

On second look I think returning a default constructed QDateTime is a problem. I've adjusted to return a default constructed value
___
*[src/persistence/history.cpp, line 847 at r1](https://reviewable.io/reviews/qtox/qtox/5828#-LorQ6GR4FManMO3ZI03:-LtpunsQAN9Rg5j4VKN7:b-896fix) ([raw file](https://github.com/qtox/qtox/blob/bc609554a8c13c7a63d683dccb3663298dde309c/src/persistence/history.cpp#L847)):*
<details><summary><i>Previously, anthonybilinski (Anthony Bilinski) wrote…</i></summary><blockquote>

I know that this is an existing Warning log, but could you bump it to Critical and add an assert? It's effectively a security issue that is narrowly being avoided if it gets to this point.
</blockquote></details>

Done.

**Comment by sphaerophoria** [comment]
PR updated. I'm starting to hit this again working on DB upgrades, hopefully it's sufficient to get it in now :)

**Review by sudden6 — APPROVED** [review]
:lgtm_strong: from my side, waiting for @anthonybilinski to lgtm and merge too

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5828#-:-Ltq2SpN0g_RdS0tG95E:bjew0ki)* status: :shipit: complete! 1 of 1 LGTMs obtained

**Comment by sudden6** [comment]
uh, wait, GH shows a merge conflict with ´history.cpp`, please fix that too :)

**Review by anthonybilinski — APPROVED** [review]
:lgtm_strong:

> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/qtox/qtox/5828#-:-Lv5IOem1_g2xFiiicsA:b-jbsf8z)* status: :shipit: complete! 2 of 1 LGTMs obtained

**Comment by anthonybilinski** [comment]
I rebased to v1.17-dev tip and tested, looks good. I like the additional generichcatform fix, I was hitting that just today :P

**Comment by anthonybilinski** [comment]
Waiting on travis despite the linked travis job having finished successfully.. going to close/re-open to see if it will fix the check.

**Comment by anthonybilinski** [comment]
urg, I'll try pushing a dummy commit then removing it.

**Comment by anthonybilinski** [comment]
closing/opening, changing base branch back and forth, and pushing then removing a dummy commit all leave travis in this state of waiting for a job's results when the job has already succeeded =/

**Pull request merged**
